README.md
# flutter_libphonenumber

A wrapper around libphonenumber with added functionality merged from the following libs:
* https://github.com/nashfive/phone_number
* https://github.com/caseyryan/flutter_multi_formatter

Uses the following native libraries:

| Platform | Library        | Version |
|----------|----------------|---------|
| Android  | libphonenumber | 8.12.52 |
| iOS      | PhoneNumberKit | 3.3.4   |

The main advantage to this lib is it lets you optionally format a phone number synchronously without making calls into libphonenumber with platform calls.

![AsYouType real-time formatting](https://media.giphy.com/media/XHk6PTxbJ5wRW6ChDz/source.gif)

![Format and parse](https://media.giphy.com/media/XGgnYYeo2YS7elAPRQ/source.gif)


## Getting Started
First you need to call the `init` function. This will load all of the available regions available on the device from libphonenumber to build a formatting mask for each country using its example number from libphonenumber.

If you don't run the init function then `formatNumberSync` will simply return the same thing passed into it without formatting anything as there won't be any masks to utilize.

We use the same approach from `flutter_multi_formatter` for masking but instead of statically defining all of our country masks, we pull them on the fly from libphonenubmer so that the masks will be automatically maintained over time.

You can either do this during your app init before calling `RunApp`, or with a `FutureBuilder` as the example app demonstrates.

```dart
await FlutterLibphonenumber().init();
```

## Formatting a number synchronously
Normally calls to libphonenumber's format function are asynchronous, and you might not want
your UI rebuilding every time you need to format a phone number.

To get around this, we load a mask of every supported phone region's example number from libphonenumber during the `init()` call. We can then use this mask to format an e164 phone number **synchronously** like this:
```dart
final rawNumber = '+14145556666';
final formattedNumber = FlutterLibphonenumber().formatNumberSync(rawNumber); // +1 414-555-6666
```

## CountryManager
When you call init, this lib will store a list of the countries and phone metadata with the following class:
```dart
class CountryWithPhoneCode {
  final String phoneCode                            // '44',
  final String countryCode                          // 'GB',
  final String exampleNumberMobileNational          // '07400 123456',
  final String exampleNumberFixedLineNational       // '0121 234 5678',
  final String phoneMaskMobileNational              // '00000 000000',
  final String phoneMaskFixedLineNational           // '0000 000 0000',
  final String exampleNumberMobileInternational     // '+44 7400 123456',
  final String exampleNumberFixedLineInternational  // '+44 121 234 5678',
  final String phoneMaskMobileInternational         // '+00 0000 000000',
  final String phoneMaskFixedLineInternational      // '+00 000 000 0000',
  final String countryName                          // 'United Kingdom';
}
```

To access this list of countries you can get at it like this:
```dart
final countries = CountryManager().countries; // List<CountryWithPhoneCode>
```

# API Reference
Here is a reference of all of the available functions.

### `Future<void> init({Map<String, CountryWithPhoneCode> overrides})`
Must be called before we can format anything. This loads all available countries on the device and calls `getAllSupportedRegions()` to then cross-reference and combine everything and save a `List<CountryWithPhoneCode>` of every available country with its phone code / mask.

Optionally provide a map of overrides where the key is the country code (ex: `GB` or `US`) and the value is a `CountryWithPhoneCode` object that should replace the data pulled from libphonenumber. This is useful if you want to customize the mask data for a given country.

### `Future<Map<String, CountryWithPhoneCode>> getAllSupportedRegions()`
Returns all of the available regions on the device in a map with each key as the region code, and the value as a `CountryWithPhoneCode` object containing all of the mobile and landline phone masks / example numbers for national / international format, phone code, region code, and country name.

Example response:
```dart
{
  "UK": [CountryWithPhoneCode()],
  "US": [CountryWithPhoneCode()]
}
```

### `Future<Map<String, String>> format(String phone, String region)`
Formats a number using libphonenumber. Under the hood this is using the AsYouType formatter. Depending on your result, you might want to use `formatNumberSync()` to utilize the phone number masks.

Will return the parsed / formatted number like this:
```dart
{
    formatted: "1 (414) 444-4444",
}
```

### `Future<Map<String, dynamic>> parse(String phone, {String region})`
Parses a number and if it is full and complete, returns some metadata associated with the number. The number must be a valid and compelte e164 formatted number to be considered valid.

Will throw an error if the number isn't valid.

Example response:
```dart
{
    country_code: 49,
    e164: '+4930123123123',
    national: '030 123 123 123',
    type: 'mobile',
    international: '+49 30 123 123 123',
    region_code: 'DE',
    national_number: '030123123123',
}
```

### `String formatNumberSync(String number, {CountryWithPhoneCode? country, PhoneNumberType phoneNumberType = PhoneNumberType.mobile, phoneNumberFormat = PhoneNumberFormat.international, bool removeCountryCodeFromResult = false, bool inputContainsCountryCode = true})`
Format a number synchronously using masks to format it. Must have ran the `init()` function to pre-populate the mask data or else the original `phone` value will be returned.

Optionally specify the phone number type to format it as (`mobile` vs `fixedLine`). This is useful when a country has more than one phone number format and you want to format it to either fit the fixed line or mobile pattern.

Use `removeCountryCodeFromResult` to remove the country code from the formatted result. Set `inputContainsCountryCode` accordingly based on if the inputted number to format contains the country code or not.

Example response:
```dart
"1 (414) 444-4444"
```

### `Future<FormatPhoneResult?> getFormattedParseResult(String phoneNumber,CountryWithPhoneCode country, {PhoneNumberType phoneNumberType = PhoneNumberType.mobile, PhoneNumberFormat phoneNumberFormat = PhoneNumberFormat.international,})`
Asynchronously formats a phone number with libphonenumber. Will return the formatted number and if it's a valid/complete number, will return the e164 value as well in the `e164` field. Uses libphonenumber's parse function to verify if it's a valid number or not. This is useful if you want to format a number and also check if it's valid, in one step.

Optionally pass a `PhoneNumberType` to format the number using either the mobile (default) or fixed line mask.

`e164` will be null if the number is not valid.

```dart
class FormatPhoneResult {
  String formattedNumber; // 1 (414) 444-4444
  String e164; // +14144444444
}
```

### `TextInputFormatter LibPhonenumberTextFormatter(...)`
* `required String country`
You must provide the country used to format the text accurately.

The text formatter also takes 5 optional arguments:
* `PhoneNumberType? phoneNumberType` specify whether to format the phone number in the mobile format using the mask for mobile numbers, or the fixed line format. Can either be `PhoneNumberType.mobile` or `PhoneNumberType.fixedLine`. Defaults to `PhoneNumberType.mobile`.

* `PhoneNumberFormat? phoneNumberFormat` specify to format using the national or international format. Can either be `PhoneNumberFormat.international` or `PhoneNumberFormat.national`. Defaults to `PhoneNumberFormat.international`.

* `FutureOr Function(String val) onFormatFinished` Optionally get a notification at this callback of the final formatted value once all formatting is completed. This is useful if you want to do something else that is triggered after formatting is done.

* `bool? inputContainsCountryCode` When true, mask will be applied assuming the input contains a country code in it.


* `int? additionalDigits` You can tell the formatter to allow additional digits on the end of the mask. This is useful for some countries which have a similar mask but varying length of numbers. It's safe to set this to a value like 3 or 5 and you won't have to think about it again.


To use it, simply add `LibPhonenumberTextFormatter()` to your `TextField`'s `inputFormatters` list:
```dart
TextField(inputFormatters: [LibPhonenumberTextFormatter(
  country: 'GB',
)])
```


CHANGELOG.md
## [1.4.0] - 2023.03.07
- Fixed `onFormatFinished` callback not being called when `shouldKeepCursorAtEndOfInput=true`.


## [1.3.0] - 2022.12.05
- Adds a new field (region_code) to the result of the parse method.
- Upgrades the dependencies:
  * libphonenumber to 8.12.52
  * PhoneNumberKit to 3.3.4
- Fixes some lint issues.
- Upgrade gradle/kotlin versions in android project.

## [1.2.4] - 2022.03.09
- Fixed some edge cases with cursor position when `shouldKeepCursorAtEndOfInput` is set to `false`

## [1.2.3] - 2022.03.08
- New optional setting `shouldKeepCursorAtEndOfInput` on `LibPhonenumberTextFormatter` which will either keep the cursor at the end of the input if the middle of the input is changed, or the cursor will remain at the same position as where it was edited.

## [1.2.2] - 2022.01.12
- Replaced jcenter with maven central (JeremyLWright).

## [1.2.1] - 2021.09.11
- Fixed bug where region was not passed to `parse` call in `getFormattedParseResult`.

# 1.2.0
- BREAKING: renamed `hideCountryCode` to `inputContainsCountryCode` in `LibPhonenumberTextFormatter`.
- BREAKING: renamed `removeCountryCode` to `removeCountryCodeFromResult` and added new parameter `inputContainsCountryCode` in `formatNumberSync()`. Use this to accurately describe if the input number contains a country code and whether or not the result should strip that country code out.

# 1.1.0
- Improvements to how masking is performed.
- Now allows for additional digits to be added on the end of the input mask for countries with varying number patterns.

# 1.0.4
- Fix number parsing issue by not attaching a leading '+' anymore. This fixes the problem parsing GB numbers starting with 07 which didn't have the 44 country code at the start.
- Bumped libphonenumber version on android.

# 1.0.3
- Nullsafety for main version.
- UK numbers will correctly remove leading 0 now on international numbers.

# 1.0.2-nullsafety
- Bumped underlying native lib versions. PhoneNumberKit -> 3.3.3, libphonenumber -> 8.12.17.

# 1.0.1-nullsafety
- Add constraint for flutter v1.10.0.

# 1.0.0-nullsafety
- Migrate to null safety.

# 0.3.11
- Update iOS minimum deployment to 9.0.

# 0.3.10
- Bumped PhoneNumberKit to 3.3 (iOS) and libphonenumber to 8.12.11 (Android).
- Fixed overflows in example app on smaller devices.

# 0.3.9
- Bump android compileSdkVersion to 29.

# 0.3.8
- Downgrade Android minSdkVersion version from 21 to 18 and bump libphonenumber version from 8.12.5 to 8.12.10 via [#3](https://github.com/bottlepay/flutter_libphonenumber/pull/3).

# 0.3.7
- Fixed bug where device locale was not correctly detected on iOS.

# 0.3.6
- Hide debug printing.

## 0.3.5
- Added ability to override country mask/phone data. Added fix for UK international numbers when someone pastes in a national format.

## 0.3.4
- Fixed bug when formatting the very first number where it wouldn't move the text selection to the very end.

## 0.3.3
- Fixed bug where realtime formatter didn't ignore leading country code if present when overrideSkipCountryCode was provided.

## 0.3.2
- Fixed formatParsePhonenumberAsync to return the correct phone number international/national format based on what was requested.

## 0.3.1
- Fixes to documentation.

## 0.3.0
- Can now format based on the national or international format of a country's phone number.

## 0.2.0
- Added ability to format numbers as either mobile or fixed line, while defaulting to mobile.

## 0.1.5
- Cleanup

## 0.1.4
- Fixed bugs in the way masking is applied which caused numbers to be formatted incorrectly to their mask.

## 0.1.3
- Removed print statements
- Countries list in CountryManager is now read-only outside of the lib

## 0.1.2
- Added example gifs

## 0.1.1
- Package metadata fixes

## 0.1.0
- Initial release


lib/flutter_libphonenumber.dart
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:flutter_libphonenumber/src/country_data.dart';
import 'package:flutter_libphonenumber/src/format_phone_result.dart';
import 'package:flutter_libphonenumber/src/input_formatter/phone_mask.dart';
import 'package:flutter_libphonenumber/src/phone_number_format.dart';
import 'package:flutter_libphonenumber/src/phone_number_type.dart';

export 'package:flutter_libphonenumber/src/country_data.dart';
export 'package:flutter_libphonenumber/src/format_phone_result.dart';
export 'package:flutter_libphonenumber/src/input_formatter/input_formatter.dart';
export 'package:flutter_libphonenumber/src/phone_number_format.dart';
export 'package:flutter_libphonenumber/src/phone_number_type.dart';

class FlutterLibphonenumber {
  FlutterLibphonenumber._internal();
  factory FlutterLibphonenumber() => _instance;
  static final FlutterLibphonenumber _instance =
      FlutterLibphonenumber._internal();

  /// Method channel
  final _channel = const MethodChannel('flutter_libphonenumber');

  /// Must call this before anything else so the countries data is populated.
  ///
  /// Optionally provide a map of overrides where the key is the country code
  /// (ex: `GB` or `US`) and the value is a `CountryWithPhoneCode` object
  /// that should replace the data pulled from libphonenumber. This is useful
  /// if you want to customize the mask data for a given country.
  Future<void> init({
    Map<String, CountryWithPhoneCode> overrides = const {},
  }) async {
    return CountryManager().loadCountries(overrides: overrides);
  }

  /// Return all available regions with their country code, phone code, and formatted
  /// example number as a mask. Useful to later format phone numbers using a mask.
  ///
  /// The response will be a [CountryWithPhoneCode]:
  ///
  /// There are some performance considerations for this so you might want to cache the
  /// result and re-use it elsewhere. There's a lot of data to iterate over.
  Future<Map<String, CountryWithPhoneCode>> getAllSupportedRegions() async {
    /// Here is what will return from the platform call:
    /// ```
    /// {
    ///   "UK": {
    ///     "phoneCode": 44,
    ///     "exampleNumberMobileNational": "07400 123456",
    ///     "exampleNumberFixedLineNational": "0121 234 5678",
    ///     "phoneMaskMobileNational": "+00 00000 000000",
    ///     "phoneMaskFixedLineNational": "+00 0000 000 0000",
    ///     "exampleNumberMobileInternational": "+44 7400 123456",
    ///     "exampleNumberFixedLineInternational": "+44 121 234 5678",
    ///     "phoneMaskMobileInternational": "+00 +00 0000 000000",
    ///     "phoneMaskFixedLineInternational": "+00 +00 000 000 0000",
    ///     "countryName": "United Kingdom"
    ///   }
    /// }
    /// ```
    final result = await _channel
            .invokeMapMethod<String, dynamic>('get_all_supported_regions') ??
        {};

    final returnMap = <String, CountryWithPhoneCode>{};
    result.forEach(
      (k, v) => returnMap[k] = CountryWithPhoneCode(
        countryName: v['countryName'] ?? '',
        phoneCode: v['phoneCode'] ?? '',
        countryCode: k,
        exampleNumberMobileNational: v['exampleNumberMobileNational'] ?? '',
        exampleNumberFixedLineNational:
            v['exampleNumberFixedLineNational'] ?? '',
        phoneMaskMobileNational: v['phoneMaskMobileNational'] ?? '',
        phoneMaskFixedLineNational: v['phoneMaskFixedLineNational'] ?? '',
        exampleNumberMobileInternational:
            v['exampleNumberMobileInternational'] ?? '',
        exampleNumberFixedLineInternational:
            v['exampleNumberFixedLineInternational'] ?? '',
        phoneMaskMobileInternational: v['phoneMaskMobileInternational'] ?? '',
        phoneMaskFixedLineInternational:
            v['phoneMaskFixedLineInternational'] ?? '',
      ),
    );
    return returnMap;
  }

  /// Formats a phone number using platform libphonenumber. Will return the parsed number.
  ///
  /// Example response:
  /// ```
  /// {
  ///   formatted: "1 (414) 444-4444",
  /// }
  /// ```
  Future<Map<String, String>> format(String phone, String region) async {
    return await _channel.invokeMapMethod<String, String>('format', {
          'phone': phone,
          'region': region,
        }) ??
        <String, String>{};
  }

  /// Parse a single string and return a map in the format below. Throws an error if the
  /// number is not a valid e164 phone number.
  ///
  /// Given a passed [phone] like '+4930123123123', the response will be:
  /// ```
  /// {
  ///   country_code: 49,
  ///   e164: '+4930123123123',
  ///   national: '030 123 123 123',
  ///   type: 'mobile',
  ///   international: '+49 30 123 123 123',
  ///   national_number: '030123123123',
  /// }
  /// ```
  Future<Map<String, dynamic>> parse(String phone, {String? region}) async {
    return await _channel.invokeMapMethod<String, dynamic>('parse', {
          'phone': phone,
          'region': region,
        }) ??
        <String, dynamic>{};
  }

  /// Given a phone number, format it automatically using the masks we have from
  /// libphonenumber's example numbers. Optionally override the country (instead
  /// of auto-detecting), the number type, format.
  ///
  /// Use [removeCountryCodeFromResult] to strip the country code from the result
  /// and set [inputContainsCountryCode] based on if the input contains a country
  /// code or not so the correct mask can be used.
  String formatNumberSync(
    String number, {
    CountryWithPhoneCode? country,
    PhoneNumberType phoneNumberType = PhoneNumberType.mobile,
    PhoneNumberFormat phoneNumberFormat = PhoneNumberFormat.international,
    bool removeCountryCodeFromResult = false,
    bool inputContainsCountryCode = true,
  }) {
    final guessedCountry =
        country ?? CountryWithPhoneCode.getCountryDataByPhone(number);

    if (guessedCountry == null) {
      return number;
    }

    var formatResult = PhoneMask(
      guessedCountry.getPhoneMask(
        format: phoneNumberFormat,
        type: phoneNumberType,
        removeCountryCodeFromMask: !inputContainsCountryCode,
      ),
    ).apply(number);

    /// Remove the country code from the result if the user set removeCountryCodeFromResult=true.
    /// Take a substring of the phone code length + 2 to account for leading `+` and space between
    /// country code and the number.
    if (removeCountryCodeFromResult && inputContainsCountryCode) {
      formatResult =
          formatResult.substring(guessedCountry.phoneCode.length + 2);
    }

    return formatResult;
  }

  /// Asynchronously formats a number, returning the e164 and the number's requested format
  /// result by specifying a [PhoneNumberType] and [PhoneNumberFormat].
  ///
  /// If the number is invalid or cannot be parsed, it will return a null result.
  Future<FormatPhoneResult?> getFormattedParseResult(
    String phoneNumber,
    CountryWithPhoneCode country, {
    PhoneNumberType phoneNumberType = PhoneNumberType.mobile,
    PhoneNumberFormat phoneNumberFormat = PhoneNumberFormat.international,
  }) async {
    try {
      /// Try to parse the number and get our result
      final res = await parse(
        phoneNumber,
        region: country.countryCode,
      );

      late final String formattedNumber;
      if (phoneNumberFormat == PhoneNumberFormat.international) {
        formattedNumber = res['international'] ?? '';
      } else if (phoneNumberFormat == PhoneNumberFormat.national) {
        formattedNumber = res['national'] ?? '';
      } else {
        /// Should never happen
        formattedNumber = '';
      }

      /// Now construct the return value based on the requested format/type.
      return FormatPhoneResult(
        e164: res['e164'] ?? '',
        formattedNumber: formattedNumber,
      );
    } catch (e) {
      // print(e);
    }

    return null;
  }
}
